include/swift/AST/ASTContext.h

```diff
@@ -290,11 +292,26 @@ class ASTContext final {
   /// Cache of autodiff-associated vector spaces.
   llvm::DenseMap<Type, Optional<VectorSpace>> AutoDiffVectorSpaces;
 
-  /// Cache of `@differentiable` attributes keyed by parameter indices. This
-  /// helps us diagnose multiple `@differentiable`s that are with respect to the
-  /// same set of parameters.
+  /// Cache of `@differentiable` attributes keyed by parameter indices. Used to
+  /// diagnose duplicate `@differentiable` attributes for the same key.
+  // NOTE(TF-680): relaxing the uniqueness condition to use derivative generic
+  // signature as a key is possible. It requires derivative generic signature
+  // mangling to avoid name collisions for SIL derivative functions with the
+  // same parameter indices but different derivative generic signatures.
   llvm::DenseMap<std::pair<Decl *, IndexSubset *>, DifferentiableAttr *>
       DifferentiableAttrs;
+
+  /// Cache of `@derivative` attributes keyed by parameter indices and
+  /// derivative function kind. Used to diagnose duplicate `@derivative`
+  /// attributes for the same key.
+  // NOTE(TF-680): relaxing the uniqueness condition to use derivative generic
+  // signature as a key is possible. It requires derivative generic signature
+  // mangling to avoid name collisions for SIL derivative functions with the
+  // same parameter indices but different derivative generic signatures.
+  llvm::DenseMap<
+      std::tuple<Decl *, IndexSubset *, AutoDiffDerivativeFunctionKind>,
+      DerivativeAttr *>
+      DerivativeAttrs;
```

Comment:
Would it be possible to use the derivative config list that you created in the previous serialization PR, so that we do not have to maintain another list?

Could the derivative config list actually also replace `DifferentiableAttrs`?

Reply:
I wondered the same. The derivative function configuration doesn't currently store whether the configuration came from a `@differentiable` attribute or `@derivative` attribute, and that information is important for users of `DifferentiableAttrs ` and `DerivativeAttrs` (checking duplicate attributes of a specific kind, not just configurations).

I haven't thought deeply about this. I'll file an issue tracking this question if it isn't resolved by the time this PR is merged.